File: nostr-knowledge/libraries/kotlin/examples/basic_usage.kt
/**
 * Nostr Kotlin Basic Usage Example
 *
 * Demonstrates key generation, event creation, relay connection,
 * publishing, and subscribing using the rust-nostr Kotlin bindings.
 *
 * Dependencies (build.gradle.kts):
 *   implementation("org.rust-nostr:nostr-sdk:0.39.0")  // Android
 *   // or: implementation("org.rust-nostr:nostr-sdk-jvm:0.39.0")  // JVM
 *
 * For nostr-java alternative, see the second section below.
 */

package com.example.nostr

import rust.nostr.sdk.*

// =============================================================================
// Section 1: Using rust-nostr Kotlin Bindings
// =============================================================================

suspend fun main() {
    println("=== Nostr Kotlin Example (rust-nostr bindings) ===\n")

    // -------------------------------------------------------------------------
    // 1. Key Generation
    // -------------------------------------------------------------------------

    // Generate a new random keypair
    val keys = Keys.generate()
    println("Public key (hex):  ${keys.publicKey().toHex()}")
    println("Public key (npub): ${keys.publicKey().toBech32()}")
    println("Secret key (nsec): ${keys.secretKey().toBech32()}")

    // Restore from an existing nsec or hex private key
    // val existingKeys = Keys.parse("nsec1...")
    // val existingKeys = Keys.parse("hex_private_key_here")

    // Derive public key from a secret key
    // val secretKey = SecretKey.parse("nsec1...")
    // val derivedKeys = Keys(secretKey)

    println()

    // -------------------------------------------------------------------------
    // 2. Event Creation and Signing
    // -------------------------------------------------------------------------

    // Create a simple text note (kind 1)
    val textNote = EventBuilder.textNote("Hello Nostr from Kotlin!")
        .signWithKeys(keys)

    println("Event ID:   ${textNote.id().toBech32()}")
    println("Event JSON: ${textNote.asJson()}")
    println()

    // Create a text note with tags
    val taggedNote = EventBuilder.textNote("Check out #nostr development in #kotlin")
        .tags(listOf(
            Tag.hashtag("nostr"),
            Tag.hashtag("kotlin")
        ))
        .signWithKeys(keys)

    println("Tagged event: ${taggedNote.asJson()}")
    println()

    // Create metadata event (kind 0)
    val metadata = Metadata()
        .setName("KotlinDev")
        .setAbout("Building nostr clients in Kotlin")
        .setNip05("kotlindev@example.com")

    val metadataEvent = EventBuilder.metadata(metadata)
        .signWithKeys(keys)

    println("Metadata event kind: ${metadataEvent.kind().asU16()}")
    println()

    // -------------------------------------------------------------------------
    // 3. Relay Connection and Management
    // -------------------------------------------------------------------------

    val client = Client()

    // Add relays
    client.addRelay("wss://relay.damus.io")
    client.addRelay("wss://nos.lol")
    client.addRelay("wss://relay.nostr.band")

    // Connect to all added relays
    println("Connecting to relays...")
    client.connect()

    // Wait briefly for connections to establish
    // In production, use proper connection state callbacks
    Thread.sleep(2000)

    println("Connected to relays.")
    println()

    // -------------------------------------------------------------------------
    // 4. Publishing Events
    // -------------------------------------------------------------------------

    // Publish the text note to all connected relays
    println("Publishing text note...")
    val sendOutput = client.sendEvent(textNote)
    println("Event sent. ID: ${sendOutput.id().toBech32()}")

    // Publish metadata
    println("Publishing metadata...")
    client.sendEvent(metadataEvent)
    println("Metadata published.")
    println()

    // Publish using the builder directly (client signs with its signer)
    // This requires setting up the client with a signer:
    // val client = Client(NostrSigner.keys(keys))
    // client.sendEventBuilder(EventBuilder.textNote("Signed by client signer"))

    // -------------------------------------------------------------------------
    // 5. Subscribing to Events
    // -------------------------------------------------------------------------

    // Build filters
    val textNoteFilter = Filter()
        .kind(Kind(1u))        // kind 1 = text notes
        .limit(10u)            // last 10 events

    // Filter for a specific author
    val authorFilter = Filter()
        .kind(Kind(1u))
        .author(keys.publicKey())
        .limit(5u)

    // Filter by hashtag
    val hashtagFilter = Filter()
        .kind(Kind(1u))
        .hashtag("nostr")
        .limit(20u)

    // Subscribe with a single filter
    println("Subscribing to text notes...")
    client.subscribe(listOf(textNoteFilter), null)

    // -------------------------------------------------------------------------
    // 6. Handling Received Events
    // -------------------------------------------------------------------------

    println("Listening for events (10 seconds)...\n")

    // Process events using the notification handler
    client.handleNotifications(object : HandleNotification {
        override fun handleEvent(relayUrl: String, subscriptionId: String, event: Event) {
            println("--- Event from $relayUrl ---")
            println("  ID:      ${event.id().toBech32()}")
            println("  Author:  ${event.author().toBech32()}")
            println("  Kind:    ${event.kind().asU16()}")
            println("  Content: ${event.content().take(100)}")
            println("  Created: ${event.createdAt().asSecs()}")
            println()
        }

        override fun handleMsg(relayUrl: String, message: RelayMessage) {
            // Handle other relay messages (NOTICE, OK, EOSE, etc.)
            println("Message from $relayUrl: $message")
        }
    })

    // -------------------------------------------------------------------------
    // 7. Advanced: NIP-19 Encoding/Decoding
    // -------------------------------------------------------------------------

    // Encode to nprofile (includes relay hints)
    val nprofile = Nip19Profile(
        keys.publicKey(),
        listOf("wss://relay.damus.io")
    )
    println("nprofile: ${nprofile.toBech32()}")

    // Encode event to nevent
    val nevent = Nip19Event(
        textNote.id(),
        keys.publicKey(),
        null,
        listOf("wss://relay.damus.io")
    )
    println("nevent: ${nevent.toBech32()}")

    // -------------------------------------------------------------------------
    // 8. Cleanup
    // -------------------------------------------------------------------------

    client.disconnect()
    println("\nDisconnected. Done.")
}


// =============================================================================
// Section 2: Alternative -- Using nostr-java from Kotlin
// =============================================================================

/**
 * nostr-java example (requires Java 21+).
 *
 * Dependencies (Maven):
 *   <dependency>
 *       <groupId>nostr-java</groupId>
 *       <artifactId>nostr-java-client</artifactId>
 *       <version>2.0.0</version>
 *   </dependency>
 */

/*
import nostr.java.client.NostrRelayClient
import nostr.java.event.GenericEvent
import nostr.java.event.GenericTag
import nostr.java.event.Kinds
import nostr.java.event.message.EventMessage
import nostr.java.identity.Identity

fun nostrJavaExample() {
    // Generate identity
    val identity = Identity.generateRandomIdentity()
    println("Public key: ${identity.publicKey}")

    // Create and sign an event
    val event = GenericEvent.builder()
        .pubKey(identity.publicKey)
        .kind(Kinds.TEXT_NOTE)
        .content("Hello from nostr-java in Kotlin!")
        .tags(listOf(
            GenericTag.of("t", "nostr-java"),
            GenericTag.of("t", "kotlin")
        ))
        .build()

    identity.sign(event)

    // Publish to relay
    NostrRelayClient("wss://relay.damus.io").use { client ->
        client.connectAsync().join()
        client.sendAsync(EventMessage(event)).join()
        println("Event sent: ${event.id}")
    }

    // Subscribe to events
    NostrRelayClient("wss://relay.damus.io").use { client ->
        client.connectAsync().join()

        val filter = EventFilter.builder()
            .kinds(listOf(Kinds.TEXT_NOTE))
            .limit(10)
            .build()

        client.subscribeAsync(filter) { receivedEvent ->
            println("Received: ${receivedEvent.content}")
        }.join()
    }
}
*/


// =============================================================================
// Section 3: Android-Specific Patterns
// =============================================================================

/**
 * Android ViewModel pattern for Nostr (conceptual -- requires Android dependencies).
 *
 * This shows how to integrate Nostr into an Android app with proper lifecycle
 * management and reactive UI updates.
 */

/*
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

data class NostrUiState(
    val events: List<Event> = emptyList(),
    val isConnected: Boolean = false,
    val error: String? = null
)

class NostrViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(NostrUiState())
    val uiState: StateFlow<NostrUiState> = _uiState

    private val client = Client()

    init {
        viewModelScope.launch {
            try {
                client.addRelay("wss://relay.damus.io")
                client.addRelay("wss://nos.lol")
                client.connect()
                _uiState.value = _uiState.value.copy(isConnected = true)
                subscribeToFeed()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    private fun subscribeToFeed() {
        val filter = Filter().kind(Kind(1u)).limit(50u)
        client.subscribe(listOf(filter), null)

        client.handleNotifications(object : HandleNotification {
            override fun handleEvent(relayUrl: String, subscriptionId: String, event: Event) {
                val current = _uiState.value.events.toMutableList()
                current.add(0, event)
                _uiState.value = _uiState.value.copy(events = current.take(100))
            }

            override fun handleMsg(relayUrl: String, message: RelayMessage) {}
        })
    }

    fun publishNote(content: String, keys: Keys) {
        viewModelScope.launch {
            val event = EventBuilder.textNote(content).signWithKeys(keys)
            client.sendEvent(event)
        }
    }

    override fun onCleared() {
        super.onCleared()
        client.disconnect()
    }
}
*/

File: nostr-knowledge/libraries/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
basic_usage.py -- Complete pynostr example

Demonstrates:
  1. Key generation (or import)
  2. Event creation and signing
  3. Connecting to relays
  4. Publishing an event
  5. Subscribing to events with filters
  6. Receiving and processing events

Requirements:
    pip install pynostr[websocket-client]

Usage:
    python basic_usage.py
"""

import time
import sys

from pynostr.key import PrivateKey
from pynostr.event import Event, EventKind
from pynostr.relay_manager import RelayManager
from pynostr.filters import FiltersList, Filters


# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

RELAYS = [
    "wss://relay.damus.io",
    "wss://nos.lol",
    "wss://relay.nostr.band",
]

# Set to None to generate a fresh key pair each run.
# Or paste your nsec here to reuse an identity:
NSEC = None  # e.g., "nsec1..."


# ---------------------------------------------------------------------------
# 1. Key Generation
# ---------------------------------------------------------------------------

def get_keys():
    """Generate a new key pair or import from NSEC."""
    if NSEC:
        private_key = PrivateKey.from_nsec(NSEC)
        print(f"Imported existing identity")
    else:
        private_key = PrivateKey()
        print(f"Generated new identity")

    public_key = private_key.public_key

    print(f"  nsec: {private_key.bech32()}")
    print(f"  npub: {public_key.bech32()}")
    print(f"  hex pubkey: {public_key.hex()}")
    print()

    return private_key, public_key


# ---------------------------------------------------------------------------
# 2. Event Creation and Signing
# ---------------------------------------------------------------------------

def create_text_note(private_key, content):
    """Create a kind-1 text note, sign it, and return the Event."""
    event = Event(
        content=content,
        kind=EventKind.TEXT_NOTE,
    )
    event.sign(private_key.hex())

    print(f"Created event:")
    print(f"  id:      {event.id}")
    print(f"  pubkey:  {event.public_key}")
    print(f"  kind:    {event.kind}")
    print(f"  content: {event.content}")
    print(f"  sig:     {event.signature[:32]}...")
    print(f"  valid:   {event.verify()}")
    print()

    return event


# ---------------------------------------------------------------------------
# 3. Relay Connection and Publishing
# ---------------------------------------------------------------------------

def publish_event(event, relays):
    """Connect to relays and publish an event."""
    relay_manager = RelayManager()

    for relay_url in relays:
        relay_manager.add_relay(relay_url)
        print(f"Added relay: {relay_url}")

    # Queue the event for publishing
    relay_manager.publish_event(event)

    # Open connections, send the event, process initial responses
    print("\nConnecting and publishing...")
    relay_manager.run_sync()

    # Wait for relay acknowledgments
    time.sleep(3)

    # Check for OK notices (NIP-20 command results)
    ok_count = 0
    while relay_manager.message_pool.has_ok_notices():
        ok_msg = relay_manager.message_pool.get_ok_notice()
        status = "accepted" if ok_msg.ok else "rejected"
        print(f"  Relay {status}: event {ok_msg.event_id[:16]}...")
        ok_count += 1

    # Check for error notices
    while relay_manager.message_pool.has_notices():
        notice = relay_manager.message_pool.get_notice()
        print(f"  Relay notice: {notice.content}")

    print(f"\nPublish complete ({ok_count} relay acknowledgments)")
    print()

    return relay_manager


# ---------------------------------------------------------------------------
# 4. Subscribing and Receiving Events
# ---------------------------------------------------------------------------

def subscribe_and_receive(relays, authors=None, limit=10):
    """
    Subscribe to text notes from specific authors (or global if None),
    receive events, and print them.
    """
    relay_manager = RelayManager()

    for relay_url in relays:
        relay_manager.add_relay(relay_url)

    # Build filter
    filter_kwargs = {
        "kinds": [EventKind.TEXT_NOTE],
        "limit": limit,
    }
    if authors:
        filter_kwargs["authors"] = authors

    filters = FiltersList([Filters(**filter_kwargs)])

    # Create subscription
    subscription_id = "basic-usage-sub"
    relay_manager.add_subscription_on_all_relays(subscription_id, filters)

    print(f"Subscribing to text notes (limit={limit})...")
    if authors:
        print(f"  Filtering by {len(authors)} author(s)")

    # Connect and fetch
    relay_manager.run_sync()

    # Allow time for events to arrive
    time.sleep(5)

    # Process received events
    events = []
    while relay_manager.message_pool.has_events():
        event_msg = relay_manager.message_pool.get_event()
        events.append(event_msg.event)

    # Check for End of Stored Events (NIP-15)
    while relay_manager.message_pool.has_eose_notices():
        eose = relay_manager.message_pool.get_eose_notice()
        print(f"  EOSE received for subscription: {eose.subscription_id}")

    print(f"\nReceived {len(events)} events:\n")

    for i, evt in enumerate(events, 1):
        # Truncate long content for display
        content_preview = evt.content[:120]
        if len(evt.content) > 120:
            content_preview += "..."

        print(f"  [{i}] Event {evt.id[:16]}...")
        print(f"      Author: {evt.public_key[:16]}...")
        print(f"      Time:   {evt.created_at}")
        print(f"      Content: {content_preview}")
        print()

    # Close subscription
    relay_manager.close_subscription_on_all_relays(subscription_id)
    print("Subscription closed.")
    print()

    return events


# ---------------------------------------------------------------------------
# 5. Complete Workflow: Reply to an Event
# ---------------------------------------------------------------------------

def reply_to_event(private_key, original_event, reply_text, relays):
    """Create and publish a reply to an existing event."""
    reply = Event(
        content=reply_text,
        kind=EventKind.TEXT_NOTE,
    )

    # NIP-10: add references to the original event and author
    reply.add_event_ref(original_event.id)
    reply.add_pubkey_ref(original_event.public_key)

    reply.sign(private_key.hex())

    print(f"Created reply:")
    print(f"  Replying to: {original_event.id[:16]}...")
    print(f"  Content: {reply_text}")
    print(f"  Tags: {reply.tags}")
    print()

    # Publish the reply
    publish_event(reply, relays)

    return reply


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    print("=" * 60)
    print("pynostr Basic Usage Example")
    print("=" * 60)
    print()

    # Step 1: Keys
    print("--- Step 1: Key Generation ---")
    private_key, public_key = get_keys()

    # Step 2: Create an event
    print("--- Step 2: Create and Sign an Event ---")
    event = create_text_note(
        private_key,
        "Hello NOSTR from pynostr! This is a test note."
    )

    # Step 3: Publish
    print("--- Step 3: Publish to Relays ---")
    publish_event(event, RELAYS)

    # Step 4: Subscribe and receive
    print("--- Step 4: Subscribe to Recent Notes ---")
    received = subscribe_and_receive(RELAYS, limit=5)

    # Step 5: Reply (only if we received events)
    if received:
        print("--- Step 5: Reply to First Received Event ---")
        reply_to_event(
            private_key,
            received[0],
            "Interesting note! (automated reply from pynostr example)",
            RELAYS,
        )

    print("=" * 60)
    print("Done.")
    print("=" * 60)


if __name__ == "__main__":
    main()
